libs/wire-api/src/Wire/API/User.hs

```diff
@@ -588,6 +596,103 @@ instance (res ~ RegisterInternalResponses) => AsUnion res (Either RegisterError
+urefToExternalId = fmap CI.original . SAML.shortShowNameID . view SAML.uidSubject
+
+urefToEmail :: SAML.UserRef -> Maybe Email
+urefToEmail uref = case uref ^. SAML.uidSubject . SAML.nameID of
+  SAML.UNameIDEmail email -> parseEmail . SAMLEmail.render . CI.original $ email
+  _ -> Nothing
+
+data CreateUserSparError
+  = CreateUserSparHandleError ChangeHandleError
+  | CreateUserSparRegistrationError RegisterError
+  deriving (Show, Generic)
+
+type CreateUserSparErrorResponses =
+  RegisterErrorResponses .++ ChangeHandleErrorResponses
+
+type CreateUserSparResponses =
+  CreateUserSparErrorResponses
+    .++ '[ WithHeaders
+             '[ DescHeader "Set-Cookie" "Cookie" Web.SetCookie,
+                DescHeader "Location" "UserId" UserId
+              ]
+             RegisterSuccess
+             (Respond 201 "User created and pending activation" SelfProfile)
+         ]
+
+type CreateUserSparInternalResponses =
+  CreateUserSparErrorResponses
+    .++ '[ WithHeaders
+             '[DescHeader "Location" "UserId" UserId]
+             SelfProfile
+             (Respond 201 "User created and pending activation" SelfProfile)
+         ]
+
+instance (res ~ CreateUserSparErrorResponses) => AsUnion res CreateUserSparError where
+  toUnion = eitherToUnion (toUnion @ChangeHandleErrorResponses) (toUnion @RegisterErrorResponses) . errToEither
+  fromUnion = errFromEither . eitherFromUnion (fromUnion @ChangeHandleErrorResponses) (fromUnion @RegisterErrorResponses)
+
+instance (res ~ CreateUserSparResponses) => AsUnion res (Either CreateUserSparError RegisterSuccess) where
+  toUnion = eitherToUnion (toUnion @CreateUserSparErrorResponses) (Z . I)
+  fromUnion = eitherFromUnion (fromUnion @CreateUserSparErrorResponses) (unI . unZ)
+
+instance (res ~ CreateUserSparInternalResponses) => AsUnion res (Either CreateUserSparError SelfProfile) where
+  toUnion = eitherToUnion (toUnion @CreateUserSparErrorResponses) (Z . I)
+  fromUnion = eitherFromUnion (fromUnion @CreateUserSparErrorResponses) (unI . unZ)
+
+errToEither :: CreateUserSparError -> Either ChangeHandleError RegisterError
+errToEither (CreateUserSparHandleError e) = Left e
+errToEither (CreateUserSparRegistrationError e) = Right e
+
+errFromEither :: Either ChangeHandleError RegisterError -> CreateUserSparError
+errFromEither (Left e) = CreateUserSparHandleError e
+errFromEither (Right e) = CreateUserSparRegistrationError e
+
+data NewUserSpar = NewUserSpar
+  { newUserSparUUID :: UUID,
+    newUserSparSSOId :: UserSSOId,
+    newUserSparDisplayName :: Name,
+    newUserSparTeamId :: TeamId,
+    newUserSparManagedBy :: ManagedBy,
+    newUserSparHandle :: Maybe Handle,
+    newUserSparRichInfo :: Maybe RichInfo
+  }
+  deriving stock (Eq, Show, Generic)
+  deriving (ToJSON, FromJSON, S.ToSchema) via (Schema NewUserSpar)
+
+instance ToSchema NewUserSpar where
+  schema =
+    object "NewUserSpar" $
+      NewUserSpar
+        <$> newUserSparUUID .= field "newUserSparUUID" genericToSchema
```

Comment:
I'm wondering if the `newUserSpar` prefix is needed for each field name of the rendered / unrendered JSON?

Reply:
Since this is an internal API used only between services we plan on merging into one soon®️, I didn't think necessary to massage the data presentation more than necessary to make it work.